### DISCUSSION TITLE
hands-on-code/hands-on-causallm-decoder

### DISCUSSION THREAD

**Discussion by giscus** [bbruceyuan/bbruceyuan.github.io#22] [Announcements]
# hands-on-code/hands-on-causallm-decoder

手写一个 Causal Language Model，或者说简化版的 transformer 中的 decoder。

https://yuanchaofa.com/hands-on-code/hands-on-causallm-decoder.html

**Comment by UltramarineJason** [comment]
你好，谢谢你的博客。 但是我想请教您以下attention mask那块为什么会有一个ifelse分支呢？ decoder的mask难道不是应该一直是一个下三角吗，用来掩盖未来的信息。在什么情况下会主动传入一个mask来自定义掩盖的信息呢？

**Reply by bbruceyuan** [reply]
## part1： 原因
先说一下为什么，我是抄的的 LLAMA 的实现（具体见下方block🤣)，

原因：**核心是为了理解 pad mask 和causal mask的区别**（但是 transformers 不同的model 有不同的实现，因此理解这里面的到底想干什么，以及这样做了和不这样做有什么区别就好了）。

## part2 ： 原理
再来说一下为什么这里有 if else，这是因为 pad_mask 和 causal LM 对应的下三角 attention 是两个不同的 mask，一个是 tokenizer 之后传过来的，告诉你哪些是 pad token，而causal attention 是告诉你最多到什么地方；
来看一下 [LLaMA 的实现](https://github.com/huggingface/transformers/blob/53fad641cfdb5105e2470bcf3ef17ea8e25cc300/src/transformers/models/llama/modeling_llama.py#L60C1-L110C23)：里面有 causal_mask 和 attention_mask 的区别，其中 attention_mask 是 pad mark，对应入参，而 `tril` or `triu` 对应 causal_mask。 

``` python
def _prepare_4d_causal_attention_mask_with_cache_position(
    attention_mask: torch.Tensor,
    sequence_length: int,
    target_length: int,
    dtype: torch.dtype,
    device: torch.device,
    min_dtype: float,
    cache_position: torch.Tensor,
    batch_size: int,
):
    """
    Creates a causal 4D mask of shape `(batch_size, 1, query_length, key_value_length)` from a 2D mask of shape
    `(batch_size, key_value_length)`, or if the input `attention_mask` is already 4D, do nothing.

    Args:
        attention_mask (`torch.Tensor`):
            A 2D attention mask of shape `(batch_size, key_value_length)` or a 4D attention mask of shape `(batch_size, 1, query_length, key_value_length)`.
        sequence_length (`int`):
            The sequence length being processed.
        target_length (`int`):
            The target length: when generating with static cache, the mask should be as long as the static cache, to account for the 0 padding, the part of the cache that is not filled yet.
        dtype (`torch.dtype`):
            The dtype to use for the 4D attention mask.
        device (`torch.device`):
            The device to plcae the 4D attention mask on.
        min_dtype (`float`):
            The minimum value representable with the dtype `dtype`.
        cache_position (`torch.Tensor`):
            Indices depicting the position of the input sequence tokens in the sequence.
        batch_size (`torch.Tensor`):
            Batch size.
    """
    if attention_mask is not None and attention_mask.dim() == 4:
        # In this case we assume that the mask comes already in inverted form and requires no inversion or slicing.
        causal_mask = attention_mask
    else:
        causal_mask = torch.full((sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device)
        if sequence_length != 1:
            causal_mask = torch.triu(causal_mask, diagonal=1)
        causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
        causal_mask = causal_mask[None, None, :, :].expand(batch_size, 1, -1, -1)
        if attention_mask is not None:
            causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
            mask_length = attention_mask.shape[-1]
            padding_mask = causal_mask[:, :, :, :mask_length] + attention_mask[:, None, None, :]
            padding_mask = padding_mask == 0
            causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
                padding_mask, min_dtype
            )

    return causal_mask
```

## part 3: 这样写真的有影响吗？
我刚刚思考了一下，好像对于结果也没有影响，假设没有 pad mask，也不过是把一些没有用的 token 做了 attention 计算。

假设一个文本长度是 2, pad 到了 4。
情况 1：只有下三角
``` python
[
[1, 0, 0, 0],
[1, 1, 0, 0],
[1, 1, 1, 0],
[1, 1, 1, 1]]
```

情况 2：有下三角 和 pad mask
``` python
[
[1, 0, 0, 0],
[1, 1, 0, 0],
[0, 0, 0, 0],
[0, 0, 0, 0]]
```

欢迎指正～

**Comment by Indeni-q** [comment]
第十六行存的self.dropout没有在后续使用上 第三十一行和第三十五行的self.layernorm_ffn = nn.LayerNorm(hidden_dim, eps=0.00001)重复了 期待发哥更新

**Reply by bbruceyuan** [reply]
非常感谢你的指正，这两个错误我都更正了~

一定多多更新

**Comment by EvilSwordman** [comment]
请问一下推理时prefill阶段为什么需要计算所有token的q。预测next token的话不是只需要计算上一个token的q就好了么？因此只需要算prefill最后一个token的q就行了？

**Reply by bbruceyuan** [reply]
这是一个非常非常非常好的问题👍

核心原因是 Decoder 中有很多层，假设只有一层，你的思考是没有问题。但是 decoder 有很多 block，上一层（假设第2层）的输入是下一层的输出，下一层（第一层）的输出依赖于 qkv 的计算，那么计算第二层的 q k v的时候，如果不知道第一层的 q k v（或者output），那么第二层的 k v 就无法计算出来了。因为 prefill 之前没有缓存好每一层的 k v 是什么。

再次感叹一下，这是一个非常好的问题~
